File: 0006_zigzag_conversion.py
# https://leetcode-cn.com/problems/zigzag-conversion/

class Solution:
    def convert(self, s, numRows):
        """
        Inputs:
            s: str
            numRows: int
        Outputs:
            res: str
        """
        
        if numRows == 1:
            return s
        
        tmp =['' for i in range(numRows)]
        
        row = 0
        down = True
        for i in range(len(s)):
            tmp[row] = tmp[row] + s[i]
            if down: # read next row
                row = row + 1
            else: # read last row
                row = row - 1
                
            if row == numRows - 1:
                down = False
            if i > 0 and row == 0:
                down = True
        
        res = ''
        for i in range(numRows):
            res = res + tmp[i]
        return res

File: Cpp/0006_zigzag_conversion.cpp
// https://leetcode-cn.com/problems/zigzag-conversion/

class Solution {
public:
	string convert(string s, int numRows) {
		if (numRows == 1) {
			return s;
		}

		int idx = 0, l = 2 * numRows - 2;
		vector<string> v(numRows);
		for (int i = 0; i < s.size(); ++i) {
			if (idx < numRows - 1) {
				v[idx] = v[idx] + s[i];
				idx = idx + 1;
			}
			else if (idx < l) {
				v[l - idx] = v[l - idx] + s[i];
				idx = idx + 1;
			}
			else {
				v[0] = v[0] + s[i];
				idx = 1;
			}
		}
		
		string res;
		for (int i = 0; i < numRows; ++i) {
			res = res + v[i];
		}
		return res;
	}
};